File: base64.py
#!/usr/bin/python

class base64:
    def greet(fx):
        def mfx(*args):
            print("Hello, I am Gumnami , nice to see you here using my tool :)")
            fx(*args)
        return mfx
    
    
    def encode(flag):
        char = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
        enc=''
        binary_string =''
        for letter in flag:
            binary_string += "{0:08b}".format(ord(letter))
        #print(binary_string)
        chunks = [binary_string[i:i+6] for i in range(0,len(binary_string),6)]
        for i in range(len(chunks)):
            if len(chunks[i]) < 6:
                #print(len(chunks[i]))
                x = 6 - len(chunks[i])
                #print(x)
                chunks[i] = "{0:06b}".format(int(chunks[i],2) << x)
        #print(chunks)
        encoded_string = ''
        for i in chunks:
            encoded_string += char[int(i,2)]
        print("here is your encoded string: " ,encoded_string)
    
    
    def decode(enc):
        char = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
        li = []
        for i in range(0,len(enc)):
            for j in range(0,len(char)):
                if enc[i] == char[j]:
                    li.append(str(j))
        print(li)
        binary = ''
        flag =''
        for num in li:
            binary += "{0:06b}".format(int(num))
        print(binary)
        chunks = [binary[i:i+8] for i in range(0,len(binary),8)]
        print(chunks)
        for i in chunks:
            flag += chr(int(i,2))
        print(flag)

    @greet
    def choice():
        a = int(input('''1> Encode a string : \n2> decode a string: \nEnter your choice: ''' ))
        match a:
            case 1:
                flag_to_encode = str(input("Enter the string to encode: "))
                base64.encode(flag_to_encode)

            case 2:
                enc_to_decode = str(input("Enter the string to decode: "))
                base64.decode(enc_to_decode)

            case _:
                print("invalid choice choice!")




base64.choice()

File: base64.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func binary(num int) string {
	var init string
	for num > 0 {
		var i int = num % 2
		val := strconv.Itoa(i)
		init += val
		num = num / 2
	}
	var result string
	for _, char := range init {
		result = string(char) + result
	}
	var l int = len(result)
	var y int = 8 - l
	if y != 0 {
		for y > 0 {
			result = string("0") + result
			y--
		}
	}
	return result
}
func debinary(num int) string {
	var init string
	for num > 0 {
		var i int = num % 2
		val := strconv.Itoa(i)
		init += val
		num = num / 2
	}
	var result string
	for _, char := range init {
		result = string(char) + result
	}
	var l int = len(result)
	var y int = 6 - l
	if y != 0 {
		for y > 0 {
			result = string("0") + result
			y--
		}
	}
	return result
}
func ascii(char rune) int {
	var val int = int(char)
	return val
}
func chunks(str string) []string {
	var x int = len(str) % 6
	for x > 0 {
		str += string("0")
		x--
	}
	var chunk = []string{}
	for i := 0; i < len(str)-6; i = i + 6 {
		chunk = append(chunk, str[i:i+6])
	}
	return chunk
}
func bintoint(chunks []string) []int64 {
	var x = []int64{}
	for _, i := range chunks {
		var y int64
		y, _ = strconv.ParseInt(i, 2, 8)
		x = append(x, y)
	}
	return x
}
func encode(chunks []int64) string {
	var arr string = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
	var ans string
	for _, i := range chunks {
		ans += string(arr[i])
	}
	return ans
}
func dechunks(str string) []string {
	var x int = len(str) % 8
	for x > 0 {
		str += string("0")
		x--
	}
	var chunk = []string{}
	for i := 0; i < len(str)-8; i = i + 8 {
		chunk = append(chunk, str[i:i+8])
	}
	return chunk
}
func findIndex(enc string) []int {
	var arr string = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
	index := []int{}
	for _, char := range enc {
		for i := range arr {
			if rune(arr[i]) == char {
				index = append(index, i)
			}
		}
	}
	return index
}
func decode(enc []int64) string {
	var str string
	for _, char := range enc {
		str += string(char)
	}
	return str
}
func main() {
	fmt.Printf("Enter the string to encode or decode : ")
	reader := bufio.NewReader(os.Stdin)
	inp, _ := reader.ReadString('\n')
	fmt.Printf("Enter 0 for encode and 1 for decode :")
	input := bufio.NewReader(os.Stdin)
	choose, _ := input.ReadString('\n')
	choice := strings.TrimSpace(choose)
	switch choice {
	case "0":
		var str string
		for _, i := range inp {
			str += binary(ascii(i))
		}
		var c []string = chunks(str)
		var x []int64 = bintoint(c)
		enc := encode(x)
		fmt.Println("The encoded string : ", enc)
	case "1":
		x := findIndex(inp)
		fmt.Println(x)
		var str string
		for i := range x {
			str += debinary(x[i])
		}
		fmt.Println(str)
		y := dechunks(str)
		fmt.Println(y)
		z := bintoint(y)
		fmt.Println(z)
		fmt.Println(decode(z))
	}
}
